### ISSUE TITLE
Simplifying options of -A, -R, -T, -F, and -N

### ISSUE THREAD

**Issue by honggyukim** [namhyung/uftrace#159]
There was a discussion regarding simplifying options regarding `-A`, `-R`, `-T`, `-F`, and `-N` as follows:

@honggyukim said (https://github.com/namhyung/uftrace/issues/112#issuecomment-313281331)
> I would like to simplify some options and trigger actions. Currently, -A serves only for arguments, -R serves only for return values, and -T is for other triggering options at the entry of function. And -F can also provide some trigger actions. Sometimes, I have to provide the same function name many times to provides it to each -A, -R, -T, -F option. We may have to make those simpler way to reduce the repetition.

And @namhyung said (https://github.com/namhyung/uftrace/issues/112#issuecomment-313288376)
> @honggyukim I agree that repeating names for each option is not a great way. Maybe we can just use -T and -R for entry/exit time trigger and provides -F, -N and -A for simpler interface (and compatibility). Thoughts?

We need to make a simpler interface for general users so I create this new issue to follow this up. The main goal is to reduce the repetition of writing same function names over and over again.

**Comment by honggyukim** [comment]
I don't know if it's intentional or not, but `-T` works at entry and return in the below example:
```
$ uftrace -T malloc@arg1,retval -T main@arg1,retval a.out 
Hello World!
# DURATION    TID     FUNCTION
   1.307 us [55948] | __monstartup();
   0.857 us [55948] | __cxa_atexit();
            [55948] | main(1) {
   6.744 us [55948] |   puts();
   6.883 us [55948] |   fflush();
   7.343 us [55948] |   malloc(100000) = 0xeb4bf0;
   3.373 us [55948] |   malloc(100000) = 0xecd2a0;
  12.251 us [55948] |   memcpy();
   1.467 us [55948] |   free();
   8.443 us [55948] |   free();
  51.374 us [55948] | } = 0; /* main */
```
From the example above, I think that why don't we just use `-T` for universal trigger for both entry and exit?

I thought that `-T` is for entry and `-R` for exit trigger as @namhyung said. But I think the opposite of `-R` is `-A`, not `-T` and it can be confusing general users.

We can just use `-R` only for displaying return value and `-A` for arguments. I think those seem to be different categories from `-T`.

To sum up, I think the purpose of each option has to be as follows:
- `-F`: child filter
- `-N`: no trace filter
- `-A`: argument option
- `-R`: return value option
- `-T`: universal trigger for both entry and exit

`-T` has to understand each meaning of syntax. I think the purpose of return trigger is for `read` trigger so in this case, I would like to suggest introducing `read-in` and `read-out` triggers as follows:
```
$ uftrace -T main@arg1,retval,read-in=proc/statm,read-out=proc/statm a.out 
Hello World!
# DURATION    TID     FUNCTION
   1.387 us [56273] | __monstartup();
   0.887 us [56273] | __cxa_atexit();
            [56273] | main(1) {
            [56273] |   /* read:proc/statm (size=27368KB, rss=3092KB, shared=2828KB) */
   8.334 us [56273] |   puts();
   7.537 us [56273] |   fflush();
   8.023 us [56273] |   malloc();
   3.180 us [56273] |   malloc();
  12.254 us [56273] |   memcpy();
   1.227 us [56273] |   free();
   9.377 us [56273] |   free();
            [56273] |   /* read:proc/statm (size=27368KB, rss=3092KB, shared=2828KB) */
  80.145 us [56273] | } = 0; /* main */
```
I would like to hear your comments. Thanks!

**Comment by honggyukim** [comment]
I could also add one more `read-inout` trigger as well to make it shorter.
```
$ uftrace -T main@arg1,retval,read-inout=proc/statm a.out 
Hello World!
# DURATION    TID     FUNCTION
   1.387 us [56273] | __monstartup();
   0.887 us [56273] | __cxa_atexit();
            [56273] | main(1) {
            [56273] |   /* read:proc/statm (size=27368KB, rss=3092KB, shared=2828KB) */
   8.334 us [56273] |   puts();
   7.537 us [56273] |   fflush();
   8.023 us [56273] |   malloc();
   3.180 us [56273] |   malloc();
  12.254 us [56273] |   memcpy();
   1.227 us [56273] |   free();
   9.377 us [56273] |   free();
            [56273] |   /* read:proc/statm (size=27368KB, rss=3092KB, shared=2828KB) */
  80.145 us [56273] | } = 0; /* main */
```

**Comment by namhyung** [comment]
Yeah I agree that we need something like "inout" triggers.

**Comment by namhyung** [comment]
Using `-T` for both entry and exit requires a way to distinguish an action to trigger when.  For example, it needs to know when `trace-off` trigger would be activated (entry or exit).  Otherwise we need to create a pair of "in" and "out" actions for each trigger.

**Comment by honggyukim** [comment]
I think `-T` can distinguish when the give trigger works using `<tr_time>` as follows:
```
              <trigger>    :=  <symbol> "@" <actions>
              <actions>    :=  <tr_action>  | <tr_action> "," <actions> 
              <tr_action>  := <action> [ <tr-time> ]
              <tr_time>    := "" | "-in" | "-out" | "-inout"
              <action>     :=  "depth="<num> | "trace" | "trace_on" | "trace_off" | "recover" |
                               "time="<time_spec> | "read="<read_spec> | "finish"
              <time_spec>  :=  <num> [ <time_unit> ]
              <time_unit>  :=  "ns" | "us" | "ms" | "s"
              <read_spec>  :=  "proc/statm" | "page-fault"
```
If `<tr_time>` is be omitted, the default is entry time trigger. And if `<tr_time>` is specified as "-in", "-out", or "-inout", the trigger can be activated at entry, exit, and both entry, exit time respectively.

**Comment by namhyung** [comment]
Sounds ok to me.  I cannot come up to a better solution for now.

**Comment by namhyung** [comment]
I pushed the code to unify trigger handling into [review/trigger-unify-v1](https://github.com/namhyung/uftrace/compare/review/trigger-unify-v1).

It doesn't support return-time triggers, but it's possible to specify argument and return value with `-T` option.  I also added 'filter' and 'notrace' action to set filters with `-T` option also.

```
$ uftrace -T main@filter,arg1,retval,read=proc/statm hello-pg
hello world
# DURATION    TID     FUNCTION
            [18941] | main(1) {
            [18941] |   /* read:proc/statm (size=28364KB, rss=4204KB, shared=3124KB) */
   2.883 us [18941] |   printf();
  31.281 us [18941] | } = 0; /* main */
```

**Comment by honggyukim** [comment]
@namhyung I'm fine with this change. Thanks for this work!

**Comment by namhyung** [comment]
Merged: 5ff0446
